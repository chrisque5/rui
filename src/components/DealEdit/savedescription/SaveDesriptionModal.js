import { useState } from 'react';
import PropTypes from 'prop-types';
import {
  Modal, Button, Input,
} from 'antd';

const { TextArea } = Input;
const SaveDesriptionModal = ({
  visible, onOk, onBack, onDescriptionChange,
}) => {
  const [isOkBtnDisabled, setIsOkBtnDisabled] = useState(true);
  const onChange = (e) => {
    const value = e.target.value;
    if (value) {
      setIsOkBtnDisabled(false);
    } else {
      setIsOkBtnDisabled(true);
    }
  };
  const onBlur = (e) => {
    onDescriptionChange(e.target.value);
  };
  return (
    <Modal
      className="save-description"
      title="Save Description"
      visible={visible}
      onOk={onOk}
      onCancel={onBack}
      bodyStyle={{ padding: '10px', minHeight: 150 }}
      footer={[
        <Button data-testid="btnBack" key="back" onClick={onBack}>Back</Button>,
        <Button data-testid="btnOk" disabled={isOkBtnDisabled} key="submit" type="primary" onClick={onOk}>Ok</Button>,
      ]}
      destroyOnClose
    >
      <TextArea rows={8} autoFocus onBlur={onBlur} onChange={onChange} name="editDescription" data-testid="editDescription" />
    </Modal>
  );
};

SaveDesriptionModal.propTypes = {
  onOk: PropTypes.func.isRequired,
  onBack: PropTypes.func.isRequired,
  onDescriptionChange: PropTypes.func.isRequired,
  visible: PropTypes.bool.isRequired,
};

export default SaveDesriptionModal;
