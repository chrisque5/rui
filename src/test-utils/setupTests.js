// The below part of the code is to fix the issue
// TypeError: window.matchMedia is not a function
Object.defineProperty(window, 'matchMedia', {
  writable: true,
  value: jest.fn().mockImplementation((query) => ({
    matches: false,
    media: query,
    onchange: null,
    addListener: jest.fn(), // deprecated
    removeListener: jest.fn(), // deprecated
    addEventListener: jest.fn(),
    removeEventListener: jest.fn(),
    dispatchEvent: jest.fn(),
  })),
});
