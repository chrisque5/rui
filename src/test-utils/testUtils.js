/* eslint-disable import/no-extraneous-dependencies, react/prop-types */
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { Form } from 'antd';

const middlewares = [thunk];
const renderWithForm = (
  ui,
  {
    initialState,
    store = configureMockStore(middlewares)(initialState),
    ...renderOptions
  } = {},
) => {
  const wrapper = ({ children }) => <Provider store={store}><Form>{children}</Form></Provider>;

  return render(ui, { wrapper, ...renderOptions });
};

const renderWithProvider = (
  ui,
  {
    initialState,
    store = configureMockStore(middlewares)(initialState),
    ...renderOptions
  } = {},
) => {
  const wrapper = ({ children }) => <Provider store={store}>{children}</Provider>;

  return render(ui, { wrapper, ...renderOptions });
};
// re-export everything
export * from '@testing-library/react';
// override render method
export { renderWithForm, renderWithProvider };
