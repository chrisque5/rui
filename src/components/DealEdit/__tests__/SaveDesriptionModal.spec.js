import {
  renderWithForm, waitFor, fireEvent,
} from 'test-utils/testUtils';
import { Simulate } from 'react-dom/test-utils';
import SaveDesriptionModal from '../savedescription/SaveDesriptionModal';

describe('<SaveDesriptionModal/>', () => {
  const defaultProps = {
    visible: true,
    onOk: jest.fn(),
    onBack: jest.fn(),
    onDescriptionChange: jest.fn(),
  };

  const renderModal = () => renderWithForm(<SaveDesriptionModal {...defaultProps} />);

  test('renders snapshot as expected', () => {
    const { container } = renderModal();
    expect(container).toMatchSnapshot();
  });

  test('verify onOk button is disabled by default', async () => {
    const { getByTestId } = renderModal();
    const okBtn = await waitFor(() => getByTestId('btnOk'));
    expect(okBtn.getAttribute('disabled')).toBe('');
  });

  test('verify on description change onOk will be enabled/disabled', async () => {
    const { getByTestId } = renderModal();
    const okBtn = await waitFor(() => getByTestId('btnOk'));
    expect(okBtn.getAttribute('disabled')).toBe('');// Disabled

    const inputText = await waitFor(() => getByTestId('editDescription'));
    fireEvent.change(inputText, { target: { value: 'edit' } });
    Simulate.blur(inputText);
    expect(okBtn.getAttribute('disabled')).toBe(null);// Enabled

    fireEvent.change(inputText, { target: { value: '' } });
    expect(okBtn.getAttribute('disabled')).toBe('');// Disabled
  });
});
