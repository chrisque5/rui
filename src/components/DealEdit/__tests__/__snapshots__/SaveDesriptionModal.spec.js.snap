// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SaveDesriptionModal/> renders snapshot as expected 1`] = `
<div>
  <form
    class="ant-form ant-form-horizontal"
  />
</div>
`;
